homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,205 @@
+        self.nanoleaf.token = self.discovery_conf.get(self.device_id, {}).get(
+            "token",  # >= 2021.4
+            self.discovery_conf.get(host, {}).get("token"),  # < 2021.4
+        )
+        if self.nanoleaf.token is not None:
+            self.context["source"] = config_entries.SOURCE_INTEGRATION_DISCOVERY
+            _LOGGER.warning(
+                "Importing Nanoleaf %s from the discovery integration", name
+            )
+            return await self.async_setup_finish()
+
+        self.context["title_placeholders"] = {"name": name}
+        return await self.async_step_link()
+
+    async def async_step_link(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf link step."""
+        if user_input is None:
+            return self.async_show_form(step_id="link")
+
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except NotAuthorizingNewTokens:
+            return self.async_show_form(
+                step_id="link", errors={"base": "not_allowing_new_tokens"}
+            )
+        except Unavailable:
+            return self.async_abort(reason="cannot_connect")
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error authorizing Nanoleaf")
+            return self.async_show_form(step_id="link", errors={"base": "unknown"})
+        return await self.async_setup_finish()
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Handle Nanoleaf configuration import."""
+        self._async_abort_entries_match({CONF_HOST: config[CONF_HOST]})
+        _LOGGER.debug(
+            "Importing Nanoleaf on %s from your configuration.yaml", config[CONF_HOST]
+        )
+        self.nanoleaf = Nanoleaf(config[CONF_HOST])
+        self.nanoleaf.token = config[CONF_TOKEN]
+        return await self.async_setup_finish()
+
+    async def async_setup_finish(self) -> FlowResult:
+        """Finish Nanoleaf config flow."""
+        try:
+            info = await self.hass.async_add_executor_job(
+                pynanoleaf_get_info, self.nanoleaf
+            )
+        except Unavailable:
+            return self.async_abort(reason="cannot_connect")
+        except InvalidToken:
+            return self.async_show_form(
+                step_id="link", errors={"base": "invalid_token"}
+            )
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Unknown error connecting with Nanoleaf at %s with token %s",
+                self.nanoleaf.host,
+                self.nanoleaf.token,
+            )
+            return self.async_abort(reason="unknown")
+        name = info["name"]
+
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: self.nanoleaf.host})
+
+        if self.context["source"] == config_entries.SOURCE_INTEGRATION_DISCOVERY:
```

Comment:
```suggestion
        if self.context["source"] in config_entries.DISCOVERY_SOURCES:
```

Not sure if `SOURCE_INTEGRATION_DISCOVERY` is the right check... still researching

Reply:
`SOURCE_INTEGRATION_DISCOVERY` is not in `DISCOVERY_SOURCES`

Reply:
I'm thinking the case where this is true `self.context["source"]` is going to be `homekit` or `zeroconf` not `integration_discovery`

Reply:
O I see you are changing it on https://github.com/home-assistant/core/pull/52199/files#diff-ea9255de1cb2d5d5a53771046f2c99ea2c5518d1bc53d51f697b89f247892de4R113

Reply:
We should not change the source.  leave it as-is and check if its a known discovery source instead